rfc-116-store-attachment-data-in-content-items.md

```diff
@@ -0,0 +1,321 @@
+
+   And these fields because the Publishing Workflow team are planning
+   to remove them from Whitehall: `order_url`, `price_in_pence`.
+
+   And these fields because they don't need to be exposed outside the
+   publishing app: `attachable_id`, `attachment_data_id`, `deleted`,
+   `ordering`, `slug`.
+
+   **Migration concerns:** This can be done without breaking
+   compatibility, as the new type will be unused for now.
+
+1. Make `id` in `attachment_asset` mandatory.
+
+    **Migration concerns:** this would need to be done in three steps:
+    1. Update travel advice publisher and manuals publisher to set an arbitrary ID for attachments.
+    2. Republish all travel advice pages and manual sections.
+    3. Make the field mandatory.
+
+1. Add a *mandatory* `details.attachments` field to all formats which
+   can have attachments.
+
+    **Migration concerns:** this would need to be done in four steps:
+    1. Add the field as optional.
+    2. Update publishing apps to write to it.
+    3. Republish all affected documents.
+    4. Make the field mandatory.
+
+1. Add a *mandatory* `details.document_attachments` list to
+   publication formats, which is an ordered list of govspeak or HTML
+   strings (using the `#/definitions/multiple_content_types` schema)
+   rendering attachments from `details.attachments`.
+
+   **Migration concerns:** this would need to be done in four steps:
+   1. Add the field as optional.
+   2. Update publishing apps (just Whitehall and Content Publisher) to write to it.
+   3. Republish all affected documents.
+   4. Make the field mandatory.
+
+1. Remove the `details.documents` field.
+
+   **Migration concerns:** this would need to be done in four steps:
+   1. Update frontend apps to render from `details.document_attachments`.
+   2. Update publishing apps to not set `details.documents`.
+   3. Republish all affected documents.
+   4. Remove the field.
+
+The final schemas will be:
+
+```
+{
+  image_asset: {
+    type: "object",
+    additionalProperties: false,
+    required: [
+      "content_type",
+      "url",
+    ],
+    properties: {
+      alt_text: { type: "string", },
+      caption: { type: "string", },
+      content_type: { type: "string", },
+      credit: { type: "string", },
+      url: { type: "string", format: "uri", },
+    },
+  },
+
+  base_attachment_asset: {
+    type: "object",
+    additionalProperties: false,
+    required: [
+      "content_type",
+      "url",
+    ],
+    properties: {
+      accessible: { type: "boolean", },
+      alternative_format_contact_email: { type: "string", },
+      content_type: { type: "string", },
+      file_size: { type: "integer", },
+      filename: { type: "string", },
+      locale: { "$ref": "#/definitions/locale", },
+      number_of_pages: { type: "integer", },
+      preview_url: { type: "string", format: "uri", },
+      title: { type: "string", },
+      url: { type: "string", format: "uri", },
+    },
+  },
+
+  attachment_asset: {
```

Comment:
It looks like we need this because Specialist Publisher will be the only app not sending IDs with its attachments. Given the relative simplicity of have a single "base" `attachment_asset`, would it be worth including Specialist Publisher in the step where we republish other attachments with arbitrary IDs? Then I think we could drop `base_attachment_asset` and just have `attachment_asset` with a mandatory ID.

There may be other considerations that are motivating this hierarchy. If so, could we articulate them in here somewhere - maybe in the 'questions' section at the bottom?

Reply:
I agree, when writing up the schema I did start to question why `attachment_asset` needed to be a separate thing.  I'll add a step about adding IDs to specialist-publisher attachments.